simulation.hpp

```diff
@@ -129,6 +129,79 @@ class Simulation
         }
     }
 
+    /** Moves hosts from one location to another
+     *
+     * @param infected Currently infected hosts
+     * @param susceptible Currently susceptible hosts
+     * @param mortality_tracker Hosts that are infected at a specific time step
+     * @param total_plants 
+     * @param step the current step of the simulation
+     * @param last_index the last index to not be used from movements
+     * @param movements a vector of ints with row_from, col_from, row_to, col_to, and num_hosts
+     * @param movement_schedule a vector matching movements with the step at which the movement from movements are applied
+     */
+    unsigned movement(IntegerRaster& infected,
+                      IntegerRaster& susceptible, 
+                      IntegerRaster& mortality_tracker,
+                      IntegerRaster& total_plants,
+                      unsigned step, unsigned last_index,
+                      const std::vector<std::vector<int>>& movements,
+                      std::vector<unsigned> movement_schedule)
+    {
+        for (unsigned i = last_index; i < movements.size(); i++) {
+            auto moved = movements[i];
+            unsigned move_schedule = movement_schedule[i];
+            if (move_schedule != step) {
+                return i;
+            }
+            int infected_moved = 0;
+            int susceptible_moved = 0;
+            int total_hosts_moved = 0;
+            double infected_mean = 0;
+            double inf_ratio = 0;
+            int row_from = moved[0];
+            int col_from = moved[1];
+            int row_to = moved[2];
+            int col_to = moved[3];
+            int hosts = moved[4];
+            if (hosts > total_plants(row_from, col_from)) {
+                total_hosts_moved = total_plants(row_from, col_from);
+            } else {
+                total_hosts_moved = hosts;
+            }
+            if (infected(row_from, col_from) > 0 && susceptible(row_from, col_from) > 0) {
+                inf_ratio = double(infected(row_from, col_from)) / double(total_plants(row_from, col_from));
+                int infected_mean = total_hosts_moved * inf_ratio;
+                if (infected_mean > 0) {
+                    std::poisson_distribution<int> distribution(infected_mean);
+                    infected_moved = distribution(generator_);
+                }
+                if (infected_moved > infected(row_from, col_from)) {
+                    infected_moved = infected(row_from, col_from);
+                } 
+                susceptible_moved = total_hosts_moved - infected_moved;
```

Comment:
Looking at it again, I think there should be one more check:

    if (infected_moved > total_hosts_moved)
        infected_moved = total_hosts_moved

Reply:
this should be taken care of with the previous check as we are assuming that infected + susceptible = total_plants and thus infected can't be larger than total hosts. That said it is probably better to be explicit here and this check shouldn't increase computational time.

Reply:
If inf_ratio is 1, then infected_moved can be easily > total_hosts_moved.

Reply:
I agree that the check is needed because even if we limit by the total available infected that does not tell anything about how many hosts we actually plan to move. We can have 50 infected hosts, but we might be moving only 20 total (so at most 20 infected).